{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "72e7f103",
   "metadata": {},
   "source": [
    "CSR SALAD Version 10.1\n",
    "8 Oct 2025\n",
    "by Jackson Cahn\n",
    "\n",
    "Adapted from:\n",
    "CSR SALAD Version 8\n",
    "11 Dec 2015\n",
    "by Jackson Cahn\n",
    "\n",
    "Copyright California Institute of Technology\n",
    "All rights reserved"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9079c6b2",
   "metadata": {},
   "source": [
    "TO use: change values in the first cell and run all. Results will appear at the bottom."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5d3af6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "infile = './my_file.pdb'\n",
    "\n",
    "max_size = 400 #maximum size of designed library, should be roughly 0.5x your screening capacity\n",
    "\n",
    "ex_motif = True #Turn on to exclude glycine-rich motif from design space. Note that even when not excluded it is deprioritized.\n",
    "ex_diphos = True #Same, for other diphosphate-contacting residues\n",
    "ex_periph = False #Same, for more peripheral residues\n",
    "\n",
    "verbose = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "47a8afa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#imports\n",
    "\n",
    "import sys,os\n",
    "from glob import glob\n",
    "from Bio.PDB import *\n",
    "from operator import itemgetter\n",
    "from numpy import array,matrix, dot,reshape\n",
    "from numpy.linalg import norm\n",
    "from math import sqrt,acos,degrees\n",
    "from itertools import product\n",
    "import pandas as pd\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2466484a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Dictionaries\n",
    "\n",
    "sidechaincenteratoms = { #the pseudocenter of each sidechain is defined as the mean xyz of these atoms\n",
    "    'GLY': ['CA'],\n",
    "    'ALA': ['CB'],\n",
    "    'VAL': ['CG1', 'CG2'],\n",
    "    'ILE': ['CG1','CG2','CD1'],\n",
    "    'LEU': ['CD1', 'CD2', 'CG', 'CB'],\n",
    "    'SER': ['OG'],\n",
    "    'THR': ['OG1', 'CG2'],\n",
    "    'ASP': ['OD1', 'OD2'],\n",
    "    'ASN': ['OD1', 'ND2'],\n",
    "    'GLU': ['OE1', 'OE2','CG'],\n",
    "    'GLN': ['OE1', 'NE2'],\n",
    "    'LYS': ['NZ', 'CE'],\n",
    "    'ARG': ['NE', 'NH1', 'NH2'],\n",
    "    'CYS': ['SG'],\n",
    "    'CSO': ['SG', 'OD'],\n",
    "    'MET': ['CE', 'SD', 'CG'],\n",
    "    'MSE': ['SE'],\n",
    "    'PHE': ['CG', 'CD1', 'CD2', 'CE1', 'CE2', 'CZ'],\n",
    "    'TYR': ['CE1', 'CE2', 'CZ', 'OH'],\n",
    "    'TRP': ['CG', 'CD1', 'CD2', 'NE1', 'CE2', 'CE3', 'CZ2', 'CZ3'],\n",
    "    'HIS': ['CG', 'ND1', 'CD2', 'CE1', 'NE2'],\n",
    "    'PRO': ['CB', 'CG', 'CD'],\n",
    "}\n",
    "\n",
    "Plibraries = { #degenerate codons for is_nadP->is_nad libraries\n",
    "    ('PHE', \"Edge\") : [['DFVY','KWC'],['DEFLVY*','KWK']],\n",
    "    ('PHE', \"Bidentate\") : [['DFVY','KWC'],['DEFLVY*','KWK'],['DEFHLQVY*','BWK']],\n",
    "    ('PHE', \"Face\") : [['F','---'],['FY','TWC'],['DFVY','KWC'],['FHLQY*','YWK']],\n",
    "    ('PHE', \"Simple\") : [['DFVY','KWC'],['DEFLVY*','KWK']],\n",
    "    ('LEU', \"Edge\") : [['ELQV','SWA']],\n",
    "    ('LEU', \"Bidentate\") : [['ELQV','SWA'],['DEHLQV','SWK']],\n",
    "    ('LEU', \"Face\") : [['FHLY','YWC'],['FILMST','WYK']],\n",
    "    ('LEU', \"Simple\") : [['ELQV','SWA'],['DEHLQV','SWK']],\n",
    "    ('ILE', \"Edge\") : [['EIKV','RWA']],\n",
    "    ('ILE', \"Bidentate\") : [['EIKV','RWA']],\n",
    "    ('ILE', \"Face\") : [['FINY','WWC'],['FHILNY','HWC']],\n",
    "    ('ILE', \"Simple\") : [['EIKV','RWA']],\n",
    "    ('MET', \"Edge\") : [['EKMV','RWG']],\n",
    "    ('MET', \"Bidentate\") : [['EKMV','RWG']],\n",
    "    ('MET', \"Face\") : [['FILM','WTK'],['FILMPST','HYK']],\n",
    "    ('MET', \"Simple\") : [['EKMV','RWG']],\n",
    "    ('VAL', \"Edge\") : [['DEV','GWK'],['ELQV','SWA'],['DEIKMNV','RWK']],\n",
    "    ('VAL', \"Bidentate\") : [['DEV','GWK']],\n",
    "    ('VAL', \"Face\") : [['DHLV','SWC'],['DHILNV','VWC']],\n",
    "    ('VAL', \"Simple\") : [['DEV','GWK']],\n",
    "    ('SER', \"Edge\") : [['INS','ADC'],['DGINSV','RDC']],\n",
    "    ('SER', \"Bidentate\") : [['DGNS','RRC'],['ADGNST','RVC']],\n",
    "    ('SER', \"Face\") : [['DGINSV','RDC']],\n",
    "    ('SER', \"Simple\") : [['DGNS','RRC'],['DGHILNRSV','VDC']],\n",
    "    ('PRO', \"Edge\") : [[\"Peripheral\",'---'],['ADHP','SMC']],\n",
    "    ('PRO', \"Bidentate\") : [['ADHP','SMC']],\n",
    "    ('PRO', \"Face\") : [['ILPT','MYA']],\n",
    "    ('PRO', \"Simple\") : [['ADHP','SMC']],\n",
    "    ('THR', \"Edge\") : [['ADNT','RMC'],['ADEKNT','RMK']],\n",
    "    ('THR', \"Bidentate\") : [['ADNT','RMC'],['ADEKNT','RMK']],\n",
    "    ('THR', \"Face\") : [['INT','AHC'],['HNPT','MMC'],['IKMNRST','ANK']],\n",
    "    ('THR', \"Simple\") : [['ADNT','RMC'],['ADEKNT','RMK']],\n",
    "    ('ALA', \"Edge\") : [['AD','GMC'],['ADNT','RMC']],\n",
    "    ('ALA', \"Bidentate\") : [['AS','KCA'],['ADGNST','RVC']],\n",
    "    ('ALA', \"Face\") : [['APT','VCA'],['AILPTV','VYA']],\n",
    "    ('ALA', \"Simple\") : [['AS','KCA'],['ADGNST','RVC']],\n",
    "    ('TYR', \"Edge\") : [['FINY','WWC'],['FHILNY','HWC']],\n",
    "    ('TYR', \"Bidentate\") : [['FINY','WWC'],['DFINVY','DWC']],\n",
    "    ('TYR', \"Face\") : [['FY','TWC'],['FHLY','YWC'],['ADFINSTVY','DHC']],\n",
    "    ('TYR', \"Simple\") : [['DEKNY*','DAK'],['DEFIKLMNVY*','DWK']],\n",
    "    ('HIS', \"Edge\") : [['DHN','VAC'],['DEHKNQ','VAK']],\n",
    "    ('HIS', \"Bidentate\") : [['DHN','VAC'],['DEHKNQ','VAK']],\n",
    "    ('HIS', \"Face\") : [['HNPT','MMC']],\n",
    "    ('HIS', \"Simple\") : [['DHN','VAC'],['DEHKNQ','VAK']],\n",
    "    ('GLN', \"Edge\") : [['Q','---'],['EQ','SAA']],\n",
    "    ('GLN', \"Bidentate\") : [['Q','---'],['EQ','SAA']],\n",
    "    ('GLN', \"Face\") : [['LPQ','CHA']],\n",
    "    ('GLN', \"Simple\") : [['Q','---'],['EQ','SAA']],\n",
    "    ('LYS', \"Edge\") : [['EK','RAA'],['EKQ','VAA'],['EIKV','RWA']], \n",
    "    ('LYS', \"Bidentate\") : [['DEGKNRS','RRK']],\n",
    "    ('LYS', \"Face\") : [['KLMQ','MWG'],['KLMPQT','MHG']],\n",
    "    ('LYS', \"Simple\") : [['DEKN','RAK'],['DEIKMNV','RWK'],['DEFIKLMNVY*','DWK']],\n",
    "    ('ASP', \"Edge\") : [['D','---'],['DE','GAK'],['DEHQ','SAK']],\n",
    "    ('ASP', \"Bidentate\") : [['D','---'],['DE','GAK'],['DEHQ','SAK']],\n",
    "    ('ASP', \"Face\") : [['D','---'],['DE','GAK'],['DEHQ','SAK']],\n",
    "    ('ASP', \"Simple\") : [['D','---'],['DE','GAK'],['DEHQ','SAK']],\n",
    "    ('GLU', \"Edge\") : [['E','---'],['EKQ','VAA']],\n",
    "    ('GLU', \"Bidentate\") : [['E','---'],['EKQ','VAA']],\n",
    "    ('GLU', \"Face\") : [['E','---'],['EKQ','VAA']],\n",
    "    ('GLU', \"Simple\") : [['E','---'],['EKQ','VAA']],\n",
    "    ('CYS', \"Edge\") : [['CDGY','KRC'],['CDFGVY','KDC']],\n",
    "    ('CYS', \"Bidentate\") : [['CDGY','KRC']],\n",
    "    ('CYS', \"Face\") : [['CFGV','KKC'],['CDFGVY','KDC']],\n",
    "    ('CYS', \"Simple\") : [['CDGY','KRC']],\n",
    "    ('TRP', \"Edge\") : [['LMRW','WKG'],['LQRW*','YDG']],\n",
    "    ('TRP', \"Bidentate\") : [['QRW*','YRG'],['CHQRWY*','YRK']],\n",
    "    ('TRP', \"Face\") : [['CFLW','TKK']],\n",
    "    ('TRP', \"Simple\") : [['QRW*','YRG'],['CHQRWY*','YRK']],\n",
    "    ('ARG', \"Edge\") : [['EGQR','SRA'],['EGLQRV','SDA']],\n",
    "    ('ARG', \"Bidentate\") : [['DEGHQR','SRK']],\n",
    "    ('ARG', \"Face\") : [['CHPRSY','YVC'],['CHNPRSTY','HVC'],['CFHILNPRSTY','HNC']],\n",
    "    ('ARG', \"Simple\") : [['DEGHQR','SRK'],['DEGKNRS','RRK'],['DEGHKNQRS','VRK']],\n",
    "    ('GLY', \"Edge\") : [['DG','GRC'],['DGV','GDC']],\n",
    "    ('GLY', \"Bidentate\") : [['DG','GRC'],['ADGV','GNC']],\n",
    "    ('GLY', \"Face\") : [['DGINSV','RDC']],\n",
    "    ('GLY', \"Simple\") : [['DG','GRC'],['ADGV','GNC'],['DEGKNRS','RRK']],\n",
    "    ('ASN', \"Edge\") : [['N','---'],['DEKN','RAK']],\n",
    "    ('ASN', \"Bidentate\") : [['DN','RAC'],['DEHKNQ','VAK']],\n",
    "    ('ASN', \"Face\") : [['FINY','WWC']], \n",
    "    ('ASN', \"Simple\") : [['DN','RAC'],['ADNT','RMC'],['ADEKNT','RMK']]\n",
    "}\n",
    "\n",
    "Nlibraries = { #degenerate codons for NAD->NADP libraries\n",
    "    ('PHE', \"Edge\") : [['F','---'],['FIKLMNY*','WWK']],\n",
    "    ('PHE', \"Bidentate\") : [['FS','TYC'],['FHLPSY','YHC']],\n",
    "    ('PHE', \"Face\") : [['CFLR','YKC'],['CFHLRY','YDC']],\n",
    "    ('PHE', \"Simple\") : [['FINSTY','WHC']],\n",
    "    ('LEU', \"Edge\") : [['EIKLQV','VWA']],\n",
    "    ('LEU', \"Bidentate\") : [['LS','TYA'],['ILST','WYA']],\n",
    "    ('LEU', \"Face\") : [['LR','CKA'],['HLR','CDC'],['HLQR','CDK']],\n",
    "    ('LEU', \"Simple\") : [['LR','CKA'],['IKLQR','MDA']],\n",
    "    ('ILE', \"Edge\") : [['IKMN','AWK']],\n",
    "    ('ILE', \"Bidentate\") : [['IKMNRS','ADK']],\n",
    "    ('ILE', \"Face\") : [['IR','AKA'],['IKRT','ANA']],\n",
    "    ('ILE', \"Simple\") : [['IKMNRS','ADK']],\n",
    "    ('MET', \"Edge\") : [['KM','AWG'],['IKMN','AWK']],\n",
    "    ('MET', \"Bidentate\") : [['KM','AWG'],['IKMNRS','ADK']],\n",
    "    ('MET', \"Face\") : [['MR','AKG']],\n",
    "    ('MET', \"Simple\") : [['KM','AWG'],['IKMNRS','ADK']],\n",
    "    ('VAL', \"Edge\") : [['AITV','RYA'],['AEIKTV','RHA']],\n",
    "    ('VAL', \"Bidentate\") : [['GISV','RKC'],['DGHILNRSV','VDC']],\n",
    "    ('VAL', \"Face\") : [['GMRV','RKG'],['AGMRTV','RBG']],\n",
    "    ('VAL', \"Simple\") : [['GISV','RKC'],['DGHILNRSV','VDC']],\n",
    "    ('SER', \"Edge\") : [['AS','KCA'],['ALSV','KYA']],\n",
    "    ('SER', \"Bidentate\") : [['GS','RGC'],['AGST','RSC']],\n",
    "    ('SER', \"Face\") : [['GS','RGC'],['AGST','RSC']],\n",
    "    ('SER', \"Simple\") : [[\"Simple\",'---'],['GS','RGC'],['AGST','RSC']],\n",
    "    ('PRO', \"Edge\") : [['LP','CYA'],['ALPV','SYA']],\n",
    "    ('PRO', \"Bidentate\") : [['PS','YCA'],['PST','HCA'],['APST','NCA']],\n",
    "    ('PRO', \"Face\") : [['PRST','MSC']],\n",
    "    ('PRO', \"Simple\") : [['PS','YCA'],['PST','HCA'],['APST','NCA']],\n",
    "    ('THR', \"Edge\") : [['AT','RCA'],['ADNT','RMC']],\n",
    "    ('THR', \"Bidentate\") : [['AST','DCA']],\n",
    "    ('THR', \"Face\") : [['AGRT','RSA'],['ADGHNPRST','VVC']],\n",
    "    ('THR', \"Simple\") : [['AST','DCA']],\n",
    "    ('ALA', \"Edge\") : [['A','---'],['AS','KCA'],['ALSV','KYA']],\n",
    "    ('ALA', \"Bidentate\") : [['AGST','RSC']],\n",
    "    ('ALA', \"Face\") : [['AGRT','RSA']],\n",
    "    ('ALA', \"Simple\") : [['AGST','RSC']],\n",
    "    ('TYR', \"Edge\") : [['FIKLMNY*','WWK']],\n",
    "    ('TYR', \"Bidentate\") : [['HNY','HAC'],['HKNQY*','HAK']],\n",
    "    ('TYR', \"Face\") : [['CHRY','YRC']],\n",
    "    ('TYR', \"Simple\") : [['HNY','HAC'],['HKNQY*','HAK']],\n",
    "    ('HIS', \"Edge\") : [['HILN','MWC']],\n",
    "    ('HIS', \"Bidentate\") : [['HNRS','MRC']],\n",
    "    ('HIS', \"Face\") : [['HNRS','MRC']],\n",
    "    ('HIS', \"Simple\") : [['HNPT','MMC'],['HNPRST','MVC']],\n",
    "    ('GLN', \"Edge\") : [['IKLQ','MWA']],\n",
    "    ('GLN', \"Bidentate\") : [['KQR','MRA'],['KPQRT','MVA']],\n",
    "    ('GLN', \"Face\") : [['KPQRT','MVA']],\n",
    "    ('GLN', \"Simple\") : [['HKNQRS','MRK']],\n",
    "    ('LYS', \"Edge\") : [['K','---'],['KN','AAK']],\n",
    "    ('LYS', \"Bidentate\") : [['KN','AAK'],['KNT','AMK'],['KNRST','AVK']],\n",
    "    ('LYS', \"Face\") : [['KNRS','ARK'],['HKNQRS','MRK']],\n",
    "    ('LYS', \"Simple\") : [['KN','AAK'],['KNT','AMK'],['KNRST','AVK']],\n",
    "    ('ASP', \"Edge\") : [['ADGNST','RVC'],['ADGINSTV','RNC']],\n",
    "    ('ASP', \"Bidentate\") : [['ADNT','RMC'],['ADGNST','RVC']],\n",
    "    ('ASP', \"Face\") : [['ADNT','RMC'],['ADHNPT','VMC']],\n",
    "    ('ASP', \"Simple\") : [['DGNS','RRC'],['ADGNST','RVC']],\n",
    "    ('GLU', \"Edge\") : [['AES*','KMA'],['AELSV*','KHA'],['AEIKLSTV*','DHA']],\n",
    "    ('GLU', \"Bidentate\") : [['EGKQR','VRA'],['DEGHKNQRS','VRW']],\n",
    "    ('GLU', \"Face\") : [['EGQR','SRA'],['AEGKPQRT','VVA']],\n",
    "    ('GLU', \"Simple\") : [['AEKST*','DMA'],['ADEKNSTY*','DMK']],\n",
    "    ('CYS', \"Edge\") : [['ACGS','KSC'],['ACFGSV','KBC']],\n",
    "    ('CYS', \"Bidentate\") : [['ACGS','KSC']],\n",
    "    ('CYS', \"Face\") : [['CHRY','YRC'],['CHNPRSTY','HVC']],\n",
    "    ('CYS', \"Simple\") : [['ACGS','KSC']],\n",
    "    ('TRP', \"Edge\") : [['LMRW','WKG']],\n",
    "    ('TRP', \"Bidentate\") : [['KRSTW*','WVG']],\n",
    "    ('TRP', \"Face\") : [['QRW*','YRG'],['CHQRWY*','YRK']],\n",
    "    ('TRP', \"Simple\") : [['KRSTW*','WVG']],\n",
    "    ('ARG', \"Edge\") : [['R','---'],['LR','CKA'],['CFHLRY','YDC']],\n",
    "    ('ARG', \"Bidentate\") : [['KNRS','ARK'],['HKNQRS','MRK']],\n",
    "    ('ARG', \"Face\") : [['R','---'],['QR','CRA'],['LPQR','CNA']],\n",
    "    ('ARG', \"Simple\") : [['KNRS','ARK'],['HKNQRS','MRK']],\n",
    "    ('GLY', \"Edge\") : [['G','---'],['AGV','GBA']],\n",
    "    ('GLY', \"Bidentate\") : [['G','---'],['AGST','RSC']],\n",
    "    ('GLY', \"Face\") : [['G','---'],['AGRT','RSA']],\n",
    "    ('GLY', \"Simple\") : [['G','---'],['AGST','RSC']],\n",
    "    ('ASN', \"Edge\") : [['NS','ARC'],['IKMN','AWK']],\n",
    "    ('ASN', \"Bidentate\") : [['NS','ARC'],['HKNQRS','MRK']],\n",
    "    ('ASN', \"Face\") : [['KNRST','AVK']],\n",
    "    ('ASN', \"Simple\") : [['NS','ARC'],['HKNQRS','MRK']],\n",
    "}\n",
    "\n",
    "degen_nucleotide_sizes = {\n",
    "    1 : ['A','G','C','T','-'],\n",
    "    2 : ['Y','R','S','W','K','M'],\n",
    "    3 : ['B','D','H','V'],\n",
    "    4 : ['N'] }\n",
    "\n",
    "degen_sizes_flat = {}\n",
    "for k in degen_nucleotide_sizes.keys():\n",
    "    for e in degen_nucleotide_sizes[k]:\n",
    "        degen_sizes_flat[e]=k\n",
    "\n",
    "_three_letter_to_one = {'CYS': 'C', 'ASP': 'D', 'SER': 'S', 'GLN': 'Q', 'LYS': 'K',\n",
    "        'ILE': 'I', 'PRO': 'P', 'THR': 'T', 'PHE': 'F', 'ASN': 'N', \n",
    "        'GLY': 'G', 'HIS': 'H', 'LEU': 'L', 'ARG': 'R', 'TRP': 'W', \n",
    "        'ALA': 'A', 'VAL': 'V', 'GLU': 'E', 'TYR': 'Y', 'MET': 'M',\n",
    "        'MSE': 'M', 'CSO': 'C'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "608546f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Analyze cofactor binding sites.\"\"\"\n",
    "\n",
    "def analyze_cofactor_binding(infile):\n",
    "    print(\"Analyzing cofactor binding site...\")\n",
    "    global LogString, passovers\n",
    "    \n",
    "    # Load and prepare structure\n",
    "    structure, cof, is_nad, ns, motif = readfile(infile)\n",
    "    phos_atoms, cof_atoms = defineatoms(structure, cof, is_nad)\n",
    "    assert len(phos_atoms) != 0, \"No phosphate atoms found\"\n",
    "    \n",
    "    # Find first shell residues\n",
    "    contactlist, recon = get_first_shell(phos_atoms, ns, motif, cof_atoms, 4.2, is_nad)\n",
    "    \n",
    "    # Expand search radius if needed\n",
    "    if len(contactlist) < 3 and not ex_periph:\n",
    "        LogString += f'Only {len(contactlist)} residues found, expanding search radius to 5.2.\\n'\n",
    "        tmp_contactlist, tmp_recon = get_first_shell(phos_atoms, ns, motif, cof_atoms, 5.2, is_nad)\n",
    "        \n",
    "        for residue in tmp_contactlist:\n",
    "            if residue.get_id()[1] % 500 not in [s.get_id()[1] % 500 for s in contactlist]:\n",
    "                recon[residue] = tmp_recon[residue]\n",
    "                LogString += f'  ->Added residue {residue}.\\n'\n",
    "    \n",
    "    # Classify residues\n",
    "    resclass = classify(recon, cof, motif, is_nad)\n",
    "    \n",
    "    # Extended edge search if needed\n",
    "    if \"Edge\" not in [e[1] for e in resclass]:\n",
    "        LogString += 'No edge residue detected, checking for weaker geometry edge residues.\\n'\n",
    "        all_residues = Selection.unfold_entities(structure[0], 'R')\n",
    "        resclass = extended_edge_search(resclass, all_residues, cof, is_nad)\n",
    "    \n",
    "    # Sort by residue number\n",
    "    resclass.sort(key=lambda c: c[0].get_id()[1] % 500)\n",
    "    assert len(resclass) != 0, \"No residues classified\"\n",
    "    \n",
    "    # Generate library\n",
    "    lib = make_library(resclass, is_nad)\n",
    "    \n",
    "    # Create library display table\n",
    "    library_df = create_library_table(resclass, lib)\n",
    "    \n",
    "    # Get additional residues for recovery\n",
    "    recovery_suggestions = get_recovery_targets(cof_atoms, ns, contactlist, structure, phos_atoms, is_nad)\n",
    "    \n",
    "    return library_df, recovery_suggestions, lib[1]\n",
    "\n",
    "def readfile(infile):\n",
    "    name = '.'.join(infile.split('.')[:-1])\n",
    "    structure = PDBParser(QUIET=True).get_structure(name, infile) # load molecule, ignore errors\n",
    "    res_list = Selection.unfold_entities(structure[0],'R')\n",
    "    motif = []\n",
    "    is_nad = False\n",
    "    global LogString\n",
    "\n",
    "    #identify cofactor molecules\n",
    "    cofactors = [] \n",
    "    for r in res_list:\n",
    "        if r.get_resname() in ['NAP' , 'NDP']:\n",
    "            cofactors.append(r)\n",
    "            LogString += 'Cofactor found: %s \\n' % str(r)\n",
    "        elif r.get_resname() in ['NAD','NAI']:\n",
    "            cofactors.append(r)\n",
    "            if is_nad == False: LogString += '**Cofactor type set to NAD**\\n'\n",
    "            is_nad = True\n",
    "            LogString += 'Cofactor found: %s \\n' % str(r)\n",
    "    if len(cofactors) == 0:\n",
    "        print(\"Cofactor not found\")\n",
    "        sys.exit()\n",
    "\n",
    "    #find glycine-rich motifs for binding pyrophosphate\n",
    "    motif_patterns = [\n",
    "        ([0, 2, 4], 5),  # G-X-G-X-G/A\n",
    "        ([0, 2, 5], 6),  # G-X-G-X-X-G/A\n",
    "        ([0, 3, 5], 6),  # G-X-X-G-X-G/A\n",
    "    ]\n",
    "\n",
    "    for i in range(len(res_list) - 6):\n",
    "        for positions, length in motif_patterns:\n",
    "            if _check_gly_motif(res_list, i, positions):\n",
    "                motif_segment = res_list[i:i + length]\n",
    "                motif.extend(motif_segment)\n",
    "                LogString += f'Gly-rich motif found: \\n--' + '\\n--'.join(str(r) for r in motif_segment) + '\\n'\n",
    "                break\n",
    "\n",
    "    # make KDTree for neighbor searching\n",
    "    atom_list = Selection.unfold_entities(structure[0],'A')\n",
    "    ns = NeighborSearch(atom_list)\n",
    "\n",
    "    return [structure,cofactors,is_nad,ns,motif]\n",
    "\n",
    "def _check_gly_motif(res_list, start, positions):\n",
    "    return bool(res_list[start + positions[0]].get_resname() == 'GLY' and\n",
    "            res_list[start + positions[1]].get_resname() == 'GLY' and\n",
    "            res_list[start + positions[2]].get_resname() in ['GLY', 'ALA'])\n",
    "\n",
    "\n",
    "def defineatoms(structure,coflist,is_nad): #returns atoms belong to the cofactor, and specifically those of the O2B (phosphate or not phosphate) moiety\n",
    "    cof_atoms = Selection.unfold_entities(coflist,'A')\n",
    "    if not is_nad: phos_names = ['O2B','P2B','O1X','O2X','O3X']\n",
    "    else: phos_names = ['O2B','N3A','C4A','O3B']\n",
    "    phos_atoms = []\n",
    "    for at in cof_atoms:\n",
    "        if at.get_name() in phos_names: phos_atoms.append(at)\n",
    "    return [phos_atoms,cof_atoms]\n",
    "\n",
    "def get_first_shell(phos_atoms, ns, motif, cof_atoms, dist, is_nad):\n",
    "    global LogString\n",
    "    \n",
    "    near_residues = {}  # {resid: (residue, vector_to_phosphate)}\n",
    "    recontacts = {}      # {resid: contact_atoms_list}\n",
    "    \n",
    "    for phos_atom in phos_atoms:\n",
    "        neighbors = _find_neighbor_atoms(phos_atom, ns, dist)\n",
    "        \n",
    "        for neighbor_atom in neighbors:\n",
    "            residue = neighbor_atom.get_parent()\n",
    "            \n",
    "            if not _is_valid_neighbor(neighbor_atom, phos_atoms, cof_atoms):\n",
    "                continue\n",
    "            \n",
    "            if is_nad and _is_behind_cofactor(neighbor_atom, phos_atom):\n",
    "                LogString += f'Atom {neighbor_atom} of residue {residue} is too far back. Skipping.\\n'\n",
    "                continue\n",
    "            \n",
    "            if _should_skip_motif_residue(residue, motif, ex_motif, passovers):\n",
    "                continue\n",
    "            \n",
    "            if not _is_residue_oriented_correctly(residue, phos_atom, is_nad):\n",
    "                continue\n",
    "            \n",
    "            # Process and store residue information\n",
    "            _process_residue_contact(\n",
    "                residue, ns, phos_atom, cof_atoms, phos_atoms, \n",
    "                dist, near_residues, recontacts, is_nad\n",
    "            )\n",
    "    \n",
    "    # Convert to output format\n",
    "    return _format_results(near_residues, recontacts)\n",
    "\n",
    "def _find_neighbor_atoms(phos_atom, ns, dist):\n",
    "    global LogString\n",
    "    center = phos_atom.get_coord()\n",
    "    neighbors = ns.search(center, dist, level='A')\n",
    "    LogString += f'Atoms near cofactor:\\n  {neighbors}\\n'\n",
    "    return neighbors\n",
    "\n",
    "\n",
    "def _is_valid_neighbor(atom, phos_atoms, cof_atoms):\n",
    "    global LogString\n",
    "    \n",
    "    if not is_aa(atom.get_parent()):\n",
    "        if atom not in phos_atoms and atom not in cof_atoms:\n",
    "            LogString += f'Atom {atom} of molecule {atom.get_parent()} is not amino acid. Skipping.\\n'\n",
    "        return False\n",
    "    \n",
    "    # Skip backbone atoms and hydrogens\n",
    "    atom_name = atom.get_name()\n",
    "    return atom_name not in ['N', 'C', 'O'] and atom_name[0] != 'H'\n",
    "\n",
    "\n",
    "def _is_behind_cofactor(neighbor_atom, phos_atom):\n",
    "    parent = phos_atom.get_parent()\n",
    "    short_dist = norm(neighbor_atom.get_vector() - parent['O2B'].get_vector())\n",
    "    long_dist = norm(neighbor_atom.get_vector() - parent['O4B'].get_vector())\n",
    "    return short_dist - long_dist > -0.75\n",
    "\n",
    "\n",
    "def _should_skip_motif_residue(residue, motif, ex_motif, passovers):\n",
    "    if residue not in motif:\n",
    "        return False\n",
    "    \n",
    "    if ex_motif:\n",
    "        if residue.get_resname() != 'GLY' and residue.get_id()[1] not in [e.get_id()[1] for e in passovers]:\n",
    "            passovers.append(residue)\n",
    "        return True\n",
    "    \n",
    "    return residue.get_resname() == 'GLY'\n",
    "\n",
    "\n",
    "def _is_residue_oriented_correctly(residue, phos_atom, is_nad):\n",
    "    global LogString\n",
    "    \n",
    "    # Special check for glycine\n",
    "    if residue.get_resname() == 'GLY':\n",
    "        ca_dist = norm(residue['CA'].get_vector() - phos_atom.get_vector())\n",
    "        n_dist = norm(residue['N'].get_vector() - phos_atom.get_vector())\n",
    "        if ca_dist > n_dist:\n",
    "            LogString += f'Residue {residue} points away from cofactor. Skipping.\\n'\n",
    "            return False\n",
    "    \n",
    "    # Check angle for residues with CA\n",
    "    if residue.has_id('CA') and 'CA' not in residue.get_resname():\n",
    "        pcenter, angle = _compute_angle(residue, phos_atom, is_nad)\n",
    "        max_angle = 100 if is_nad else 90\n",
    "        if degrees(angle) > max_angle:\n",
    "            LogString += f'Residue {residue} points away from cofactor. Skipping.\\n'\n",
    "            return False\n",
    "    \n",
    "    return True\n",
    "\n",
    "def _process_residue_contact(residue, ns, phos_atom, cof_atoms, phos_atoms, dist, near_residues, recontacts, is_nad):\n",
    "    global LogString\n",
    "    \n",
    "    resid = residue.get_id()[1] % 500\n",
    "    pcenter, _ = _compute_angle(residue, phos_atom, is_nad)\n",
    "    vector_to_phos = pcenter - phos_atom.get_vector()\n",
    "    \n",
    "    # Check for recontacts\n",
    "    contact_atoms = _check_recontacts(residue, ns, cof_atoms, phos_atoms, dist, is_nad)\n",
    "    if contact_atoms is False:\n",
    "        return\n",
    "    \n",
    "    # Important contact atoms to preserve across updates\n",
    "    important_atoms = ['O1A', 'O2A', 'O1N', 'O2N', 'O3', 'C5A', 'O3B']\n",
    "    \n",
    "    if resid not in near_residues or recontacts[resid] is False:\n",
    "        # First encounter of this residue ID\n",
    "        near_residues[resid] = (residue, vector_to_phos)\n",
    "        recontacts[resid] = contact_atoms\n",
    "        _log_residue_contacts(residue, contact_atoms)\n",
    "        \n",
    "    elif vector_to_phos.norm() < near_residues[resid][1].norm():\n",
    "        # Found closer instance - update but preserve important contacts\n",
    "        old_contacts = recontacts[resid]\n",
    "        merged_contacts = _merge_contact_lists(contact_atoms, old_contacts, important_atoms)\n",
    "        \n",
    "        near_residues[resid] = (residue, vector_to_phos)\n",
    "        recontacts[resid] = merged_contacts\n",
    "        \n",
    "    else:\n",
    "        # Not closer, but merge any new important contacts\n",
    "        old_contacts = recontacts[resid]\n",
    "        recontacts[resid] = _merge_contact_lists(old_contacts, contact_atoms, important_atoms)\n",
    "\n",
    "def _format_results(near_residues, recontacts):\n",
    "    \"\"\"Format results for return.\"\"\"\n",
    "    residue_list = []\n",
    "    contact_dict = {}\n",
    "    \n",
    "    for resid, (residue, _) in near_residues.items():\n",
    "        residue_list.append(residue)\n",
    "        if recontacts[resid] is not False:\n",
    "            contact_dict[residue] = recontacts[resid]\n",
    "    \n",
    "    return [residue_list, contact_dict]\n",
    "\n",
    "def _merge_contact_lists(primary, secondary, important_atoms):\n",
    "    global LogString\n",
    "    merged = list(primary)\n",
    "    for atom in secondary:\n",
    "        if atom not in merged and atom in important_atoms:\n",
    "            merged.append(atom)\n",
    "            LogString += f'Atom {atom} appended to contacts.\\n'\n",
    "    return merged\n",
    "\n",
    "\n",
    "def _log_residue_contacts(residue, contact_atoms):\n",
    "    \"\"\"Log residue contact information.\"\"\"\n",
    "    global LogString\n",
    "    recontacting_atoms = [atom for atom in contact_atoms if atom != 'Peripheral']\n",
    "    LogString += f'Residue {residue} contacts atoms {recontacting_atoms} of cofactor.\\n'\n",
    "    if 'Peripheral' in contact_atoms:\n",
    "        LogString += '  ->Assigned as peripheral.\\n'\n",
    "\n",
    "\n",
    "\n",
    "def _compute_angle(res, root, is_nad):\n",
    "    pcenter = _get_pseudocenter(res)\n",
    "    if res.get_resname() == 'GLY': return [pcenter,0]\n",
    "    CA = res['CA'].get_vector()\n",
    "    if not is_nad: root_loc = root.get_parent()['P2B'].get_vector()\n",
    "    else: root_loc = root.get_parent()['C2B'].get_vector()\n",
    "    v1 = pcenter - CA\n",
    "    v2 = root_loc - CA\n",
    "    cosang = dot(v1,v2)/(norm(v1)*norm(v2))\n",
    "    ang = acos(cosang)\n",
    "    return [pcenter,ang]\n",
    "\n",
    "\n",
    "def _get_pseudocenter(r):\n",
    "    if r.get_resname() in sidechaincenteratoms.keys():\n",
    "        center_atoms = sidechaincenteratoms[r.get_resname()]\n",
    "    else:\n",
    "        print(f\"{r} is a nonstandard amino acid.\")\n",
    "        return r['CA'].get_vector()\n",
    "    center_atom_coords = []\n",
    "    for at in Selection.unfold_entities(r,'A'):\n",
    "        if at.get_name() in list(center_atoms): center_atom_coords.append(at.get_vector())\n",
    "    if len(center_atom_coords) == 0:\n",
    "        print(f\"{r} is missing its sidechain.\")\n",
    "        return r['CA'].get_vector()\n",
    "    pseudocenter = center_atom_coords[0]\n",
    "    if len(center_atom_coords) != 1:\n",
    "        for v in center_atom_coords[1:]: pseudocenter = pseudocenter + v\n",
    "    pseudocenter = pseudocenter / len(center_atom_coords)\n",
    "    return pseudocenter\n",
    "\n",
    "def _check_recontacts(r,ns,cof_atoms,phos,dist,is_nad): #recontacts are the atoms of the cofactor within a certain distance of the residue\n",
    "    recontacts = []\n",
    "    if is_nad:\n",
    "        for e in phos:\n",
    "            if e.get_name() == 'O3B': \n",
    "                phos.remove(e)\n",
    "    pcount = 0\n",
    "    for at in Selection.unfold_entities(r,'A'):\n",
    "        if at.get_name() in sidechaincenteratoms[r.get_resname()]: \n",
    "            neighbors = ns.search(at.get_coord(),dist*1.025,level='A')\n",
    "            for neighbor_atom in neighbors:\n",
    "                if neighbor_atom not in cof_atoms: continue\n",
    "                if neighbor_atom in phos:\n",
    "                    pcount += 1\n",
    "                    continue\n",
    "                if neighbor_atom.get_name() not in recontacts: recontacts.append(neighbor_atom.get_name())\n",
    "    if pcount == 0:\n",
    "        if r.get_resname() in ['SER','THR','TYR','HIS','GLN','ASN','LYS','ARG','GLU','ASP']: #common H binders\n",
    "            if len(recontacts) == 0: recontacts.append(\"Peripheral\")\n",
    "        else:\n",
    "            return False\n",
    "    return recontacts\n",
    "\n",
    "\n",
    "def classify(recontacts, cofactor, motif, is_nad):\n",
    "    global LogString\n",
    "    \n",
    "    class_list = []\n",
    "    \n",
    "    for residue in recontacts.keys():\n",
    "        residue_class = _determine_residue_class(residue, recontacts[residue], cofactor, motif, is_nad)\n",
    "        \n",
    "        if residue_class:\n",
    "            # Handle exclusions for certain classes\n",
    "            if _should_exclude_residue(residue, residue_class):\n",
    "                passovers.append(residue)\n",
    "            else:\n",
    "                class_list.append([residue, residue_class])\n",
    "    \n",
    "    return class_list\n",
    "\n",
    "def _determine_residue_class(residue, recontacts, cofactor, motif, is_nad):\n",
    "    global LogString\n",
    "    \n",
    "    # Check motif membership\n",
    "    if not ex_motif and residue in motif:\n",
    "        return \"Motif\"\n",
    "    \n",
    "    # Check peripheral recontacts\n",
    "    if recontacts == [\"Peripheral\"]:\n",
    "        return \"Peripheral\"\n",
    "    \n",
    "    # Handle residues with no recontacts\n",
    "    if not recontacts:\n",
    "        return _classify_no_recontact_residue(residue, cofactor, is_nad)\n",
    "    \n",
    "    # Check face contacts (C5A plus ring atoms)\n",
    "    ring_atoms = {'C4A', 'N3A', 'C2A', 'N1A', 'C6A', 'C8A', 'N7A'}\n",
    "    if 'C5A' in recontacts and bool(ring_atoms & set(recontacts)): #bool is checking if the merged set has length > 0\n",
    "        return \"Face\"\n",
    "    \n",
    "    # Check other ring contacts with plane verification\n",
    "    back_ring_atoms = {'N9A', 'C4A', 'N3A', 'C2A', 'N1A', 'C6A', 'N7A'}\n",
    "    if bool(back_ring_atoms & set(recontacts)):\n",
    "        class_check = _plane_check(residue, cofactor, is_nad) #returns \"Edge\",\"Face\", or None\n",
    "        if class_check:\n",
    "            return class_check\n",
    "    \n",
    "    # Check diphosphate binding\n",
    "    diphosphate_atoms = {'O1A', 'O2A', 'O1N', 'O2N', 'O3'}\n",
    "    if bool(diphosphate_atoms & set(recontacts)):\n",
    "        return \"Pyrophosphate\"\n",
    "    \n",
    "    # Check bidentate binding\n",
    "    if 'O3B' in recontacts:\n",
    "        return \"Bidentate\"\n",
    "    \n",
    "    # NAD-specific ring-binder check (Search at expanded distance)\n",
    "    if is_nad and _is_ring_binder(residue):\n",
    "        return \"Ring-binder\"\n",
    "    \n",
    "    # Default classification\n",
    "    return \"Nonsimple\"\n",
    "\n",
    "def _classify_no_recontact_residue(residue, cofactor, is_nad):\n",
    "    global LogString\n",
    "    \n",
    "    if residue.get_resname() != 'ARG' or not residue.has_id('CD'):\n",
    "        return \"Simple\"\n",
    "    \n",
    "    # Check arginine CD distances to cofactor atoms\n",
    "    _, n_cof = _nearest(cofactor, residue, is_nad)\n",
    "    \n",
    "    cd_c4a_dist = norm(n_cof['C4A'].get_vector() - residue['CD'].get_vector())\n",
    "    if cd_c4a_dist < 4.2:\n",
    "        LogString += f'Arg {residue} defined as face on the basis of CD-C4A distance of {cd_c4a_dist:.1f}.\\n'\n",
    "        return \"R-chain\"\n",
    "    \n",
    "    cd_c5a_dist = norm(n_cof['C5A'].get_vector() - residue['CD'].get_vector())\n",
    "    if cd_c5a_dist < 4.2:\n",
    "        LogString += f'Arg {residue} defined as face on the basis of CD-C5A distance of {cd_c5a_dist:.1f}.\\n'\n",
    "        return \"R-chain\"\n",
    "    \n",
    "    return \"Simple\"\n",
    "\n",
    "def _nearest(cofactors, residue, is_nad): #find nearest cofactor to residue (for multimers)\n",
    "    res_pcenter = _get_pseudocenter(residue)\n",
    "    min_dist = [20,None]\n",
    "    for cof in cofactors:\n",
    "        if not is_nad: P_coord = cof['P2B'].get_vector()\n",
    "        else: P_coord = cof['O2B'].get_vector()\n",
    "        if norm(res_pcenter-P_coord) < min_dist[0]:\n",
    "            min_dist = [norm(res_pcenter-P_coord),cof] #update if find something closer\n",
    "    cofactor = min_dist[1]\n",
    "    return [res_pcenter, cofactor]\n",
    "\n",
    "\n",
    "def _is_ring_binder(residue):\n",
    "    global LogString\n",
    "    \n",
    "    expanded_contacts = []\n",
    "    for atom in Selection.unfold_entities(residue, 'A'):\n",
    "        if atom.get_name() in sidechaincenteratoms[residue.get_resname()]:\n",
    "            neighbors = ns.search(atom.get_coord(), 4.4, level='A')\n",
    "            for neighbor_atom in neighbors:\n",
    "                if neighbor_atom in cof_atoms:\n",
    "                    expanded_contacts.append(neighbor_atom.get_name())\n",
    "    \n",
    "    if 'O2B' not in expanded_contacts:\n",
    "        LogString += f'Residue {residue} is peripheral even with expanded search radius.\\n'\n",
    "        return True\n",
    "    \n",
    "    return False\n",
    "\n",
    "\n",
    "def _should_exclude_residue(residue, residue_class):\n",
    "    \"\"\"Determine if a residue should be excluded based on its class and settings.\"\"\"\n",
    "    exclusion_rules = {\n",
    "        \"Peripheral\": ex_periph,\n",
    "        \"Pyrophosphate\": ex_diphos,\n",
    "        \"Ring-binder\": ex_periph  # Ring-binder follows peripheral exclusion\n",
    "    }\n",
    "    \n",
    "    return exclusion_rules.get(residue_class, False)\n",
    "\n",
    "def _plane_check(res, cofactors, is_nad):\n",
    "    \"\"\"This function transforms the coordinates of residues into a new coordinate system based on\n",
    "    the position and orientation of the cofactor, to better determine where they are in reference\n",
    "    to the plane of the adenine ring\"\"\"\n",
    "    global LogString\n",
    "    \n",
    "    res_pcenter, cofactor = _nearest(cofactors, res, is_nad)\n",
    "    if cofactor is None:\n",
    "        return None\n",
    "    \n",
    "    # Transform to adenine-centered coordinate system\n",
    "    transform_matrix = _get_adenine_transform_matrix(cofactor)\n",
    "    pc_vec = _transform_vector(res_pcenter, cofactor['C4A'].get_vector(), transform_matrix)\n",
    "    \n",
    "    LogString += f'Residue {res} has pseudocenter-C4A vector <{pc_vec[0]:.2f},{pc_vec[1]:.2f},{pc_vec[2]:.2f}> in adenine-centered coordinate space.\\n'\n",
    "    \n",
    "    # Check for offset face positioning\n",
    "    if _is_offset_face(pc_vec):\n",
    "        LogString += '  ->Meets threshold for assignment as face.\\n'\n",
    "        return \"Offset face\"\n",
    "    \n",
    "    # Check for edge positioning\n",
    "    if _is_potential_edge(pc_vec) and res.get_resname() != 'GLY':\n",
    "        if _confirm_edge_by_ca_position(res, cofactor, transform_matrix):\n",
    "            LogString += '  ->Meets threshold for assignment as edge.\\n'\n",
    "            return \"Edge\"\n",
    "    \n",
    "    # Check planar residues for R-chain classification\n",
    "    if res.get_resname() in ['ARG', 'GLN', 'TYR', 'PHE', 'GLU', 'TRP', 'HIS', 'PRO']:\n",
    "        if _is_parallel_to_adenine(res, cofactor):\n",
    "            LogString += '  ->Below threshold, assigned as face.\\n'\n",
    "            return \"R-chain\"\n",
    "    \n",
    "    return None\n",
    "\n",
    "\n",
    "def _get_adenine_coordinate_system(cofactor):\n",
    "    c4a = cofactor['C4A'].get_vector()\n",
    "    n3a = cofactor['N3A'].get_vector()\n",
    "    c5a = cofactor['C5A'].get_vector()\n",
    "    \n",
    "    # Create orthonormal basis\n",
    "    x_axis = (n3a - c4a).normalized()\n",
    "    y_axis = (c5a - c4a).normalized()\n",
    "    z_axis = (x_axis ** y_axis).normalized()  # Normal to adenine plane\n",
    "    \n",
    "    return x_axis, y_axis, z_axis\n",
    "\n",
    "\n",
    "def _get_adenine_transform_matrix(cofactor):\n",
    "    \"\"\"Create transformation matrix to adenine-centered coordinate system.\"\"\"\n",
    "    x_axis, y_axis, z_axis = _get_adenine_coordinate_system(cofactor)\n",
    "    \n",
    "    # Build transformation matrix (inverse of basis matrix)\n",
    "    basis_matrix = matrix([\n",
    "        [x_axis[0], y_axis[0], z_axis[0]],\n",
    "        [x_axis[1], y_axis[1], z_axis[1]],\n",
    "        [x_axis[2], y_axis[2], z_axis[2]]\n",
    "    ])\n",
    "    \n",
    "    return basis_matrix.I\n",
    "\n",
    "\n",
    "def _transform_vector(point, origin, transform_matrix):\n",
    "    \"\"\"Transform a vector into the new coordinate system.\"\"\"\n",
    "    vec = point - origin\n",
    "    vec_matrix = matrix([[vec[0]], [vec[1]], [vec[2]]])\n",
    "    transformed = transform_matrix * vec_matrix\n",
    "    \n",
    "    # Convert matrix result to list\n",
    "    return [float(transformed[0, 0]), float(transformed[1, 0]), float(transformed[2, 0])]\n",
    "\n",
    "\n",
    "def _is_offset_face(pc_vec):\n",
    "    z_distance = abs(pc_vec[2])\n",
    "    xy_sum = abs(pc_vec[0]) + abs(pc_vec[1])\n",
    "    xy_bounds = (abs(pc_vec[0]) < 3 and -2 < pc_vec[1] < 3)\n",
    "    \n",
    "    # Check magnitude of transformed vector\n",
    "    vector_magnitude = sqrt(pc_vec[0]**2 + pc_vec[1]**2 + pc_vec[2]**2)\n",
    "    \n",
    "    return bool(z_distance > 2.5 and \n",
    "            (xy_sum < 2.5 or xy_bounds) and \n",
    "            vector_magnitude < 5.35)\n",
    "\n",
    "\n",
    "def _is_potential_edge(pc_vec):\n",
    "    return bool(abs(pc_vec[2]) < 3.2 and \n",
    "            abs(pc_vec[2]) < abs(pc_vec[1]) and \n",
    "            abs(pc_vec[0]) < 5)\n",
    "\n",
    "\n",
    "def _confirm_edge_by_ca_position(res, cofactor, transform_matrix):\n",
    "    \"\"\"Confirm edge classification by checking CA position.\"\"\"\n",
    "    global LogString\n",
    "    \n",
    "    ca_vec = _transform_vector(res['CA'].get_vector(), cofactor['C4A'].get_vector(), transform_matrix)\n",
    "    \n",
    "    LogString += f'Residue {res} has CA-C4A vector <{ca_vec[0]:.2f},{ca_vec[1]:.2f},{ca_vec[2]:.2f}> in adenine-centered coordinate space.\\n'\n",
    "    \n",
    "    # CA should be more in-plane than out, and on negative y side\n",
    "    return bool(abs(ca_vec[2]) < max(abs(ca_vec[0]), abs(ca_vec[1])) and \n",
    "            ca_vec[1] < 0)\n",
    "\n",
    "\n",
    "def _get_adenine_coordinate_system(cofactor):\n",
    "    \"\"\"Get the orthonormal basis vectors for the adenine ring coordinate system.\"\"\"\n",
    "    c4a = cofactor['C4A'].get_vector()\n",
    "    n3a = cofactor['N3A'].get_vector()\n",
    "    c5a = cofactor['C5A'].get_vector()\n",
    "    \n",
    "    # Create orthonormal basis\n",
    "    x_axis = (n3a - c4a).normalized()\n",
    "    y_axis = (c5a - c4a).normalized()\n",
    "    z_axis = (x_axis ** y_axis).normalized()  # Normal to adenine plane\n",
    "    \n",
    "    return x_axis, y_axis, z_axis\n",
    "\n",
    "\n",
    "def _get_adenine_transform_matrix(cofactor):\n",
    "    \"\"\"Create transformation matrix to adenine-centered coordinate system.\"\"\"\n",
    "    x_axis, y_axis, z_axis = _get_adenine_coordinate_system(cofactor)\n",
    "    \n",
    "    # Build transformation matrix (inverse of basis matrix)\n",
    "    basis_matrix = matrix([\n",
    "        [x_axis[0], y_axis[0], z_axis[0]],\n",
    "        [x_axis[1], y_axis[1], z_axis[1]],\n",
    "        [x_axis[2], y_axis[2], z_axis[2]]\n",
    "    ])\n",
    "    \n",
    "    return basis_matrix.I\n",
    "\n",
    "\n",
    "def _is_parallel_to_adenine(res, cofactor):\n",
    "    global LogString\n",
    "    \n",
    "    # Get atoms defining the residue plane\n",
    "    plane_atoms = _get_plane_atoms(res)\n",
    "    if len(plane_atoms) < 3:\n",
    "        return False\n",
    "    \n",
    "    # Calculate normal vector to residue plane\n",
    "    vectors = [res[atom].get_vector() for atom in plane_atoms[:3]]\n",
    "    x_vec = vectors[1] - vectors[0]\n",
    "    y_vec = vectors[2] - vectors[0]\n",
    "    residue_normal = (x_vec ** y_vec).normalized()\n",
    "    \n",
    "    _, _, adenine_normal = _get_adenine_coordinate_system(cofactor)\n",
    "    \n",
    "    # Check angle between normals\n",
    "    cos_angle = dot(adenine_normal, residue_normal) / (norm(adenine_normal) * norm(residue_normal))\n",
    "    angle = degrees(acos(cos_angle))\n",
    "    deviation = abs(90 - angle)\n",
    "    \n",
    "    LogString += f'Angle between normal vector to side-chain of residue {res} and normal vector to adenine is {90 - deviation:.2f}.\\n'\n",
    "    \n",
    "    return bool(deviation < 20)\n",
    "\n",
    "\n",
    "def _get_plane_atoms(res):\n",
    "    resname = res.get_resname()\n",
    "    \n",
    "    if resname in ['GLU', 'GLN']:\n",
    "        atoms = list(sidechaincenteratoms[resname])\n",
    "        atoms.append('CD')\n",
    "        return atoms\n",
    "    elif resname == 'PRO':\n",
    "        return ['CB', 'CD', 'N']\n",
    "    elif resname in sidechaincenteratoms:\n",
    "        return sidechaincenteratoms[resname][:3]\n",
    "    \n",
    "    return []\n",
    "\n",
    "def extended_edge_search(resclass, res_list, cofactors, is_nad):\n",
    "    global LogString\n",
    "    \n",
    "    # Find residues that might have floor residues before them\n",
    "    trigger_classes = {\"Offset face\", \"Face\", \"R-chain\"}\n",
    "    \n",
    "    for res, res_class in resclass:\n",
    "        if res_class not in trigger_classes:\n",
    "            continue\n",
    "            \n",
    "        prior_residues = _get_prior_residues(res, res_list, 2) # Check the two residues before this one\n",
    "        \n",
    "        for prior in prior_residues:\n",
    "            if _should_reclassify_as_floor(prior, res, resclass, cofactors, is_nad):\n",
    "                _update_classification(resclass, prior, \"Floor\")\n",
    "    \n",
    "    return resclass\n",
    "\n",
    "def _get_prior_residues(target_res, res_list, n=2):\n",
    "    for i, res in enumerate(res_list):\n",
    "        if res == target_res and i >= n:\n",
    "            return res_list[i-n:i]\n",
    "    return []\n",
    "\n",
    "\n",
    "def _should_reclassify_as_floor(candidate, reference_res, resclass, cofactors, is_nad):\n",
    "    global LogString\n",
    "    \n",
    "    # Skip glycines and things already classified as Edge-like or Face-like\n",
    "    if candidate.get_resname() == 'GLY':\n",
    "        return False\n",
    "    current_class = _get_current_class(candidate, resclass)\n",
    "    if current_class in {\"Edge\", \"Face\", \"Offset face\", \"R-chain\"}:\n",
    "        return False\n",
    "    \n",
    "    # check geometry\n",
    "    res_pcenter, cofactor = _nearest(cofactors, candidate, is_nad)\n",
    "    if cofactor is None:\n",
    "        return False\n",
    "    \n",
    "    o2_vec = cofactor['O2B'].get_vector()\n",
    "    if norm(o2_vec - res_pcenter) > norm(o2_vec - candidate['CA'].get_vector()):\n",
    "        return False\n",
    "    \n",
    "    edge_angle = _calculate_edge_angle(candidate, res_pcenter, cofactor)\n",
    "    LogString += f'Angle between residue {candidate} side chain and adenine edge is {edge_angle:.2f}.\\n'\n",
    "    if edge_angle > 45:\n",
    "        return False\n",
    "    \n",
    "    normal_angle = _calculate_normal_angle(candidate, res_pcenter, cofactor)\n",
    "    deviation = abs(90 - normal_angle)\n",
    "    LogString += f'Angle between side-chain of residue {candidate} and normal vector to adenine is {deviation:.2f}.\\n'\n",
    "    \n",
    "    if deviation < 20:\n",
    "        LogString += '  ->Meets threshold for assignment as floor residue.\\n'\n",
    "        return True\n",
    "    \n",
    "    return False\n",
    "\n",
    "def _calculate_edge_angle(residue, res_pcenter, cofactor):\n",
    "    edge_vector = cofactor['N9A'].get_vector() - cofactor['N3A'].get_vector()\n",
    "    residue_vector = res_pcenter - residue['CA'].get_vector()\n",
    "    \n",
    "    cos_angle = dot(edge_vector, residue_vector) / (norm(edge_vector) * norm(residue_vector))\n",
    "    return degrees(acos(cos_angle))\n",
    "\n",
    "\n",
    "def _calculate_normal_angle(residue, res_pcenter, cofactor):\n",
    "    edge_vector = cofactor['N9A'].get_vector() - cofactor['N3A'].get_vector()\n",
    "    plane_vector = cofactor['C4A'].get_vector() - cofactor['C5A'].get_vector()\n",
    "    normal_vector = edge_vector ** plane_vector  # Cross product\n",
    "    \n",
    "    residue_vector = res_pcenter - residue['CA'].get_vector()\n",
    "    cos_angle = dot(normal_vector, residue_vector) / (norm(normal_vector) * norm(residue_vector))\n",
    "    return degrees(acos(cos_angle))\n",
    "\n",
    "def _get_current_class(residue, resclass):\n",
    "    for res, res_class in resclass:\n",
    "        if res.get_id() == residue.get_id():\n",
    "            return res_class\n",
    "    return None\n",
    "\n",
    "def _update_classification(resclass, residue, new_class):\n",
    "    # If it already has a class\n",
    "    for entry in resclass:\n",
    "        if entry[0].get_id() == residue.get_id():\n",
    "            entry[1] = new_class\n",
    "            return    \n",
    "    # Not found, add new entry\n",
    "    resclass.append([residue, new_class])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ec4bb642",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Design library for cofactor specificity reversal\"\"\"\n",
    "\n",
    "def make_library(resinfo, is_nad):\n",
    "    print(\"Designing library for cofactor specificity reversal...\")\n",
    "    raw_lib = []\n",
    "    for res, resclass  in resinfo:\n",
    "        paired_info = (res.get_resname(),_simplify_resclass(resclass))\n",
    "        if not is_nad: appendor = (res.get_id()[1],Plibraries[paired_info])\n",
    "        else: appendor = (res.get_id()[1],Nlibraries[paired_info])\n",
    "\n",
    "        if resclass in [\"Peripheral\",\"Ring-binder\",\"Pyrophosphate\",\"Motif\"]:\n",
    "            appendor[1].insert(0,[_three_letter_to_one[res.get_resname()],'---'])\n",
    "        raw_lib.append(appendor)\n",
    "    library = select_size(raw_lib,resinfo, is_nad)\n",
    "    return library\n",
    "\n",
    "def _simplify_resclass(rc):\n",
    "    if rc in [\"Edge\",\"Bidentate\",\"Face\",\"Simple\"]: return rc\n",
    "    elif rc == \"Floor\": return \"Edge\"\n",
    "    elif rc in [\"R-chain\",\"Offset face\",\"Ring-binder\"]: return \"Face\"\n",
    "    elif rc in [\"Nonsimple\",\"Simple*\",\"Peripheral\"]: return \"Simple\"\n",
    "    elif rc in [\"Motif\",\"Pyrophosphate\"]: return \"Bidentate\"\n",
    "    else:\n",
    "        print(\"unknown resclass\", rc)\n",
    "        sys.exit()\n",
    "\n",
    "def select_size(raw_lib, resclass, is_nad):\n",
    "    global LogString\n",
    "    \n",
    "    resclass_dict = {e[0].get_id()[1]: e[1] for e in resclass}\n",
    "\n",
    "    raw_lib = _reduce_minimum_size(raw_lib, resclass, is_nad) #add omission options\n",
    "\n",
    "    index_dict = _find_target_size(raw_lib, resclass_dict) #find indices for dictionary under max size\n",
    "    \n",
    "    library_out = _format_output_library(raw_lib, resclass, index_dict)\n",
    "    result_string = f\"\\n Suggested library size: {_count_size(raw_lib, index_dict)} \"\n",
    "    \n",
    "    return [library_out, result_string]\n",
    "\n",
    "def _reduce_minimum_size(raw_lib, resclass, is_nad):\n",
    "    global LogString\n",
    "    \n",
    "    min_size = _count_size(raw_lib, [0] * len(raw_lib))\n",
    "    LogString += f'Minimum default library size is {min_size}.\\n'\n",
    "    \n",
    "    if min_size <= max_size:\n",
    "        return raw_lib\n",
    "    \n",
    "    # Priority order for amino acids to omit\n",
    "    priority_orders = {\n",
    "        'P': ['W','C','M','F','P','I','L','E','D','V','Q','H','A','Y','G','N','T','K','R','S'],\n",
    "        'N': ['C','W','M','Y','H','F','T','Q','A','P','G','R','S','N','L','V','E','K','I','D']\n",
    "    }\n",
    "    order = priority_orders['N'] if is_nad else priority_orders['P']\n",
    "    \n",
    "    for aa in order:\n",
    "        for lib_posn, res_entry in enumerate(resclass):\n",
    "            residue = res_entry[0]\n",
    "            if _three_letter_to_one[residue.get_resname()] == aa:\n",
    "                # Add option to omit this residue\n",
    "                omission_option = [aa, '---']\n",
    "                raw_lib[lib_posn] = (\n",
    "                    raw_lib[lib_posn][0], \n",
    "                    [omission_option] + raw_lib[lib_posn][1]\n",
    "                )\n",
    "                \n",
    "                min_size = _count_size(raw_lib, [0] * len(raw_lib))\n",
    "                LogString += f'\\n  ->Adding option to omit residue {raw_lib[lib_posn][0]} from library. '\n",
    "                LogString += f'Minimum size is now {min_size}.\\n'\n",
    "                \n",
    "                if min_size <= max_size:\n",
    "                    return raw_lib\n",
    "    \n",
    "    return raw_lib\n",
    "\n",
    "\n",
    "def _find_target_size(raw_lib, resclass_dict):\n",
    "    \"\"\"Reduce library size from maximum by adjusting variant indices.\"\"\"\n",
    "    global LogString\n",
    "    \n",
    "    # Start with largest possible library\n",
    "    index_dict = [len(entry[1]) - 1 for entry in raw_lib]\n",
    "    size = _count_size(raw_lib, index_dict)\n",
    "    LogString += f' Starting at largest library size, {size}.\\n'\n",
    "    \n",
    "    if size <= max_size:\n",
    "        return index_dict\n",
    "    \n",
    "    resclass_priority = [\n",
    "        \"Motif\", \"Pyrophosphate\", \"Ring-binder\", \"Peripheral\", \n",
    "        \"Floor\", \"Edge\", \"R-chain\", \"Offset face\", \"Face\", \n",
    "        \"Simple*\", \"Simple\", \"Nonsimple\", \"Bidentate\"\n",
    "    ]\n",
    "    \n",
    "    # Keep cycling through priorities until we reach target size\n",
    "    while size > max_size:\n",
    "        found_reduction = False\n",
    "        \n",
    "        for resclass_type in resclass_priority:\n",
    "            for i in range(len(raw_lib)):\n",
    "                res_id = raw_lib[i][0]\n",
    "                \n",
    "                if index_dict[i] > 0 and resclass_dict[res_id] == resclass_type:\n",
    "                    index_dict[i] -= 1\n",
    "                    size = _count_size(raw_lib, index_dict)\n",
    "                    LogString += f'Reducing library size at residue {res_id}. Library size is now {size}.\\n'\n",
    "                    found_reduction = True\n",
    "                    \n",
    "                    if size <= max_size:\n",
    "                        return index_dict\n",
    "                    break  # Move to next resclass type after one reduction\n",
    "        \n",
    "        # Safety check - if we made no reductions in a full pass, we're stuck\n",
    "        if not found_reduction:\n",
    "            LogString += f'WARNING: Cannot reduce library size further. Stuck at {size}.\\n'\n",
    "            break\n",
    "    \n",
    "    return index_dict\n",
    "\n",
    "\n",
    "def _format_output_library(raw_lib, resclass, index_dict):\n",
    "    global passovers\n",
    "    \n",
    "    # Special case for single position\n",
    "    if len(index_dict) == 1:\n",
    "        return raw_lib[0][1][index_dict[0]]\n",
    "    \n",
    "    # Build library and track omitted residues\n",
    "    library_out = []\n",
    "    for i in range(len(raw_lib)):\n",
    "        variant = raw_lib[i][1][index_dict[i]]\n",
    "        library_out.append(variant)\n",
    "        \n",
    "        # Track residues marked for omission\n",
    "        if variant[1] == '---':\n",
    "            passovers.append(resclass[i][0])\n",
    "    \n",
    "    return library_out\n",
    "\n",
    "\n",
    "def _count_size(raw_lib, index_dict):\n",
    "    \"\"\"Calculate total library size for given variant indices.\"\"\"\n",
    "    assert len(raw_lib) == len(index_dict)\n",
    "    size = 1\n",
    "    for i, idx in enumerate(index_dict):\n",
    "        codon = raw_lib[i][1][idx][1]\n",
    "        size *= _expand_codon(codon)\n",
    "    return size\n",
    "\n",
    "\n",
    "def _expand_codon(codon):\n",
    "    \"\"\"Calculate number of sequences a degenerate codon represents.\"\"\"\n",
    "    size = 1\n",
    "    for nucleotide in codon:\n",
    "        size *= degen_sizes_flat[nucleotide]\n",
    "    return size\n",
    "\n",
    "\n",
    "\n",
    "def create_library_table(resclass, lib):\n",
    "    \"\"\"Create a pandas DataFrame with the library results.\"\"\"\n",
    "    data = []\n",
    "    \n",
    "    for i, (residue, classification) in enumerate(resclass):\n",
    "        res_name = residue.get_resname()\n",
    "        res_id = residue.get_id()[1]\n",
    "        \n",
    "        if len(resclass) != 1:\n",
    "            codon = lib[0][i][1]\n",
    "            amino_acids = lib[0][i][0]\n",
    "        else:\n",
    "            codon = lib[0][1]\n",
    "            amino_acids = lib[0][0]\n",
    "        \n",
    "        data.append({\n",
    "            'Residue': f'{res_name} {res_id}',\n",
    "            'Type': classification,\n",
    "            'Codon': codon,\n",
    "            'Amino Acids': amino_acids\n",
    "        })\n",
    "    \n",
    "    return pd.DataFrame(data)\n",
    "\n",
    "# def style_results_table(df):\n",
    "#     \"\"\"Apply styling to results DataFrame.\"\"\"\n",
    "#     return df.style.set_properties(**{\n",
    "#         'text-align': 'left',\n",
    "#         'font-size': '12pt',\n",
    "#     }).set_table_styles([\n",
    "#         {'selector': 'th', 'props': [('font-weight', 'bold'), \n",
    "#                                       ('text-align', 'center'),\n",
    "#                                       ('background-color', '#f0f0f0')]}\n",
    "#     ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8ef5f520",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Identify residues for site-saturation recovery.\"\"\"\n",
    "def get_recovery_targets(cof_atoms, ns, contactlist, structure, phos_atoms, is_nad):\n",
    "    print(\"Indentifying positions for activity recovery...\")\n",
    "    global passovers\n",
    "    \n",
    "    # Get backing residues\n",
    "    backinglist = list(_get_backing(cof_atoms, ns, contactlist))\n",
    "    backinglist.sort(key=lambda c: c.get_id()[1] % 500)\n",
    "    \n",
    "    # Get hydrogen bonding residues\n",
    "    hbonders = find_hbonds(structure, contactlist, ns)\n",
    "    \n",
    "    # Get second shell residues\n",
    "    secondshell = list(get_second_shell(phos_atoms, ns, contactlist, is_nad))\n",
    "    \n",
    "    # Add unique hbonders to second shell\n",
    "    existing_ids = ([e.get_id()[1] for e in backinglist] + \n",
    "                    [e.get_id()[1] for e in secondshell])\n",
    "    for h in hbonders:\n",
    "        if h.get_id()[1] not in existing_ids:\n",
    "            secondshell.append(h)\n",
    "    \n",
    "    secondshell.sort(key=lambda c: c.get_id()[1] % 500)\n",
    "    \n",
    "    # Create recovery priority table\n",
    "    recovery_positions = []\n",
    "    \n",
    "    for res in passovers:\n",
    "        recovery_positions.append({\n",
    "            'Residue': f'{res.get_resname()} {res.get_id()[1]}',\n",
    "            'Priority': 'High',\n",
    "            'Reason': 'Excluded from library'\n",
    "        })\n",
    "    \n",
    "    for res in backinglist:\n",
    "        recovery_positions.append({\n",
    "            'Residue': f'{res.get_resname()} {res.get_id()[1]}',\n",
    "            'Priority': 'Medium',\n",
    "            'Reason': 'Backing residue'\n",
    "        })\n",
    "    \n",
    "    for res in secondshell:\n",
    "        recovery_positions.append({\n",
    "            'Residue': f'{res.get_resname()} {res.get_id()[1]}',\n",
    "            'Priority': 'Low',\n",
    "            'Reason': 'Second shell/H-bonding'\n",
    "        })\n",
    "    \n",
    "    return pd.DataFrame(recovery_positions) if recovery_positions else None\n",
    "\n",
    "def _get_backing(cof_atoms,ns,contactlist):\n",
    "    backing_residues = {}\n",
    "    for at in cof_atoms:\n",
    "        if at.get_name() in ['N7A','C5A','N6A','N1A']:\n",
    "            center = at.get_coord()\n",
    "            neighbors = ns.search(center,4.1,level='R')\n",
    "            for n in neighbors:\n",
    "                if n.has_id('CA') and 'CA' not in n.get_resname() and n.get_id()[1]%500 not in [r.get_id()[1]%500 for r in contactlist]:\n",
    "                    if n.get_id()[1]%500 not in backing_residues: backing_residues[n.get_id()[1]%500] = n\n",
    "    return backing_residues.values()\n",
    "\n",
    "def get_second_shell(phos_atoms,ns,contactlist, is_nad):\n",
    "    second_shell_residues = {}\n",
    "    for at in phos_atoms:\n",
    "        center = at.get_coord()\n",
    "        neighbors = ns.search(center, 8.2,level='R')\n",
    "        for n in neighbors:\n",
    "            if n.has_id('CA') and 'CA' not in n.get_resname() and n.get_id()[1]%500 not in [r.get_id()[1]%500 for r in contactlist]:\n",
    "                if (is_nad == False and n.get_resname() in ['LYS','ARG','HIS']) or (is_nad and n.get_resname() in ['ASP','GLU']):\n",
    "                    if n.get_id()[1]%500 not in second_shell_residues: second_shell_residues[n.get_id()[1]%500] = n \n",
    "    return second_shell_residues.values()\n",
    "\n",
    "def find_hbonds(structure,contactlist,ns):\n",
    "    hbond_residues = {}\n",
    "    res_list = Selection.unfold_entities(structure[0],'R')\n",
    "    expanded_contacts = []\n",
    "    for r in res_list:\n",
    "        if r.get_id()[1]%500 in [c.get_id()[1]%500 for c in contactlist] and r not in expanded_contacts: expanded_contacts.append(r)\n",
    "    for r in expanded_contacts:\n",
    "        res_atoms = Selection.unfold_entities(r,'A')\n",
    "        for at in res_atoms:\n",
    "            if at.get_name()[0] in ['O','N'] and at.get_name() not in ['O','N']:\n",
    "                nats = ns.search(at.get_coord(),3.8,level='A')\n",
    "                for n in nats:\n",
    "                    if n.get_name()[0] in ['O','N'] and n.get_name() not in ['O','N']: #an oxygen or nitrogen but not backbone\n",
    "                        nr = n.get_parent()\n",
    "                        if is_aa(nr) == False: continue\n",
    "                        if nr in expanded_contacts: continue\n",
    "                        if nr.get_id()[1]%500 not in hbond_residues: hbond_residues[nr.get_id()[1]%500] = nr\n",
    "    return hbond_residues.values()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3312f60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Main execution\n",
    "\n",
    "#Initialize holders\n",
    "LogString = ''\n",
    "by_aa = {key: [] for key in sidechaincenteratoms.keys()}\n",
    "lengthlist = []\n",
    "passovers = []\n",
    "\n",
    "# Main execution\n",
    "library_df, recovery_df, library_info = analyze_cofactor_binding(infile)\n",
    "\n",
    "# Display library table\n",
    "print(\"\\n=== LIBRARY DESIGN RESULTS ===\\n\")\n",
    "#styled_results = style_results_table(library_df)\n",
    "display(library_df)\n",
    "print(f\"\\n{library_info}\")\n",
    "\n",
    "# Display recovery suggestions\n",
    "if recovery_df is not None and not recovery_df.empty:\n",
    "    print(\"\\n=== SITE-SATURATION MUTAGENESIS TARGETS ===\")\n",
    "    print(\"The following residues should be targeted for activity recovery:\\n\")\n",
    "    \n",
    "    # Group by priority for cleaner display\n",
    "    for priority in ['High', 'Medium', 'Low']:\n",
    "        priority_df = recovery_df[recovery_df['Priority'] == priority]\n",
    "        if not priority_df.empty:\n",
    "            print(f\"\\n{priority} Priority:\")\n",
    "            display(priority_df[['Residue', 'Reason']])\n",
    "else:\n",
    "    print(\"\\nNo obvious amino acids for activity recovery.\")\n",
    "\n",
    "# Display log if verbose\n",
    "if verbose:\n",
    "    print(\"\\n=== ACTIVITY LOG ===\\n\")\n",
    "    # Clean up HTML tags from log\n",
    "    clean_log = LogString.replace('<', '[').replace('>', ']').replace('[br]', '\\n')\n",
    "    print(clean_log)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env1",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
